workflow-ui/backstack-android/src/androidTest/java/com/squareup/workflow1/ui/backstack/test/BackStackContainerTest.kt

```diff
@@ -0,0 +1,175 @@
+package com.squareup.workflow1.ui.backstack.test
+
+import android.view.View
+import androidx.test.espresso.Espresso.onView
+import androidx.test.espresso.assertion.ViewAssertions.matches
+import androidx.test.espresso.matcher.ViewMatchers.isCompletelyDisplayed
+import androidx.test.espresso.matcher.ViewMatchers.withTagValue
+import androidx.test.ext.junit.rules.ActivityScenarioRule
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import com.google.common.truth.Truth.assertThat
+import com.squareup.workflow1.ui.WorkflowUiExperimentalApi
+import com.squareup.workflow1.ui.backstack.test.fixtures.BackStackTestActivity
+import com.squareup.workflow1.ui.backstack.test.fixtures.BackStackTestActivity.TestRendering
+import com.squareup.workflow1.ui.backstack.test.fixtures.ViewStateTestView
+import com.squareup.workflow1.ui.backstack.test.fixtures.ViewStateTestView.ViewHooks
+import com.squareup.workflow1.ui.backstack.toBackStackScreen
+import org.hamcrest.Matchers.equalTo
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@OptIn(WorkflowUiExperimentalApi::class)
+@RunWith(AndroidJUnit4::class)
+class BackStackContainerTest {
+
+  @Rule @JvmField val scenarioRule = ActivityScenarioRule(BackStackTestActivity::class.java)
+  private val scenario get() = scenarioRule.scenario
+
+  @Test fun restores_view_on_pop_without_config_change() {
+    lateinit var firstScreen: TestRendering
+    lateinit var firstView: View
+
+    scenario.onActivity { activity ->
+      // Set some view state to be saved and restored.
+      activity.currentTestView.viewState = "hello world"
+      firstView = activity.currentTestView
+      firstScreen = activity.backstack!!.top
+    }
+
+    // Navigate to another screen.
+    setBackstack(firstScreen, TestRendering("new screen"))
+    assertThat(viewForScreen("new screen").viewState).isEqualTo("")
+
+    // Navigate back.
+    setBackstack(firstScreen)
+
+    viewForScreen("initial").let {
+      // Ensure that the view instance wasn't re-used.
+      assertThat(it).isNotSameInstanceAs(firstView)
+
+      // Check that the view state was actually restored.
+      assertThat(it.viewState).isEqualTo("hello world")
+    }
+  }
+
+  @Test fun restores_view_after_config_change() {
+    scenario.onActivity { activity ->
+      // Set some view state to be saved and restored.
+      activity.currentTestView.viewState = "hello world"
+    }
+
+    // Destroy and recreate the activity.
+    scenario.recreate()
```

Comment:
OMG.

O.
M.
G.

I didn't know this was possible. We need a lot of this in house.